optimum/habana/transformers/generation/utils.py

```diff
@@ -383,16 +495,32 @@ def generate(
             )
 
         if generation_config.static_shapes:
-            # token_idx is the current index in the generation process, it is incremented each time a new token is generated
-            model_kwargs["token_idx"] = torch.tensor(inputs_tensor.shape[-1], device=inputs_tensor.device)
             # Pad inputs to have static shapes during generation, this gives better performance than dynamic shapes on HPUs
-            inputs_tensor = torch.nn.functional.pad(
-                inputs_tensor, (0, generation_config.max_new_tokens), value=generation_config.pad_token_id
-            )
-            if model_kwargs["attention_mask"] is not None:
-                model_kwargs["attention_mask"] = torch.nn.functional.pad(
-                    model_kwargs["attention_mask"], (0, generation_config.max_new_tokens), value=0
+            # In encoder_decoder models, Inputs are already padded
+            if not self.config.is_encoder_decoder:
+                # token_idx is the current index in the generation process, it is incremented each time a new token is generated
+                model_kwargs["token_idx"] = torch.tensor(inputs_tensor.shape[-1], device=inputs_tensor.device)
+                inputs_tensor = torch.nn.functional.pad(
+                    inputs_tensor, (0, generation_config.max_length), value=generation_config.pad_token_id
                 )
+                if model_kwargs["attention_mask"] is not None:
+                    model_kwargs["attention_mask"] = torch.nn.functional.pad(
+                        model_kwargs["attention_mask"], (0, generation_config.max_length), value=0
+                    )
+            else:
+                model_kwargs["token_idx"] = torch.tensor(1, device=inputs_tensor.device)
+                if model_kwargs.get("decoder_attention_mask", None) is None and generation_config.use_cache:
+                    model_kwargs["decoder_attention_mask"] = self._prepare_decoder_attention_mask(
+                        generation_config.max_length,
+                        inputs_tensor.shape[0],
+                        generation_config.pad_token_id,
+                        inputs_tensor.device,
+                    )
+
+            # if model_kwargs.get("past_key_values", None) is None and generation_config.use_cache:
+            #     model_kwargs["past_key_values"] = self._prepare_past_key_values(
+            #         generation_config.max_length, inputs_tensor.shape[0], inputs_tensor.device
+            #     )
```

Comment:
To remove?
Also, do we still need `_prepare_past_key_values` if this is commented? I didn't find any other place where it's called.